src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,138 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final int ELEMENT_NOT_FOUND = -1;
+    private static final double CAPACITY_MULTIPLIER = 1.5;
+
+    private int size;
+    private T[] innerArray;
+
+    @SuppressWarnings("unchecked")
+    public ArrayList() {
+        this.size = 0;
+        this.innerArray = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
+        if (size + 1 >= innerArray.length) {
+            addCapacity();
+        }
 
+        innerArray[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexValidity(index, true);
+
+        if (size + 1 >= innerArray.length) {
+            addCapacity();
+        }
 
+        moveElementsForward(index, innerArray);
+        innerArray[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            this.add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexValidity(index);
+
+        return innerArray[index];
     }
 
     @Override
     public void set(T value, int index) {
+        checkIndexValidity(index);
 
+        innerArray[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexValidity(index);
+
+        T elementToRemove = innerArray[index];
+        moveElementsBackwards(index, innerArray);
```

Comment:
Use System.arraycopy() to move your array elements. It is more efficient and clean compared to manual element shifting.